scipy/optimize/bfgsh.py

```diff
@@ -0,0 +1,427 @@
+    if x0.ndim == 0:
+        x0.shape = (1,)
+
+    gfk = fprime(x0)
+
+    k,N = 0,len(x0)
+    I = numpy.eye(N, dtype=int)
+
+    # Initialize inv Hess as Identity matrix
+    Hk = I
+
+    # Store your old func_max
+    if f is not None:
+        old_fval = f(x0, *args)
+
+    xk = x0
+
+    # Criteria on gradient for continuing simulation
+    gnorm = vecnorm(gfk, ord=norm)
+    
+    warnflag = 0
+    if retall:
+        allvecs = [x0]
+    # Main Loop:
+    if disp:
+        print("Alpha, Beta, H_reset = %lg, %lg, %s"
+                   % (alpha,beta,str(H_reset)))
+    while (gnorm > gtol) and (k < maxiter):
+        if disp:
+            print("Step %d, " % k),
+        # Get your step direction
+        sk = -numpy.dot(Hk, gfk)
+
+        # If we are doing unreasonably small step sizes, quit
+        if numpy.linalg.norm(sk*alpha) < 1E-7:
+            if disp:
+                print("Error - Step size unreasonable (%lg)" 
+                            % numpy.linalg.norm(sk*alpha))
+            warnflag = 2
+            break
+
+        # Hold new parameters
+        xkp1 = xk + alpha * sk
+
+        # Get the new gradient
+        gfkp1 = fprime(xkp1)
+
+        # Check if max has increased
+        if f is not None:
+            fval = f(xkp1, *args)
+        if f is not None and fval > old_fval:
+            # Step taken overstepped the minimum.  Lowering step size
+            if disp:
+                print("\tResetting System as %lg > %lg!"
+                        % (fval, old_fval))
+                print("\talpha: %lg" % alpha),
+            alpha *= float(beta)
+            if disp:
+                print("-> %lg\n" % alpha)
+
+            # Reset the Inverse Hessian if desired - This is recommended!
+            if H_reset:
+                Hk = I
+            continue
+        
+        # Store new parameters, as it has passed the check
+        # (fval < old_fval is True)
+        xk = xkp1
+
+        # Store for output if desired
+        if retall:
+            allvecs.append(xk)
+        
+        # Store new max value in old_max for future comparison
+        if f is not None:
+            old_fval = fval
+
+        # Get difference in gradients for further calculations
+        yk = gfkp1 - gfk
+        # Store new gradient in old gradient
+        gfk = gfkp1
+
+        # Update the conditional check
+        gnorm = vecnorm(gfk, ord=norm)
+
+        if disp:
+            print("gnorm %lg" % gnorm)
+
+        # If callback is desired
+        if callback is not None:
+            callback(xk)
+
+        # Increment the loop counter
+        k += 1
+        gnorm = vecnorm(gfk, ord=norm)
+        if (gnorm <= gtol):
+            break
+
+        # Run BFGS Update for the Inverse Hessian
+        A1 = I - sk[:, numpy.newaxis] * yk[numpy.newaxis, :]
+        A2 = I - yk[:, numpy.newaxis] * sk[numpy.newaxis, :]
+        Hk = numpy.dot(A1, numpy.dot(Hk, A2)) + \
+             (sk[:, numpy.newaxis] * sk[numpy.newaxis, :])
```

Comment:
Yes. I had to remove the y^T * s part (or as the original code had it:

    rhok = 1.0 / (numpy.dot(yk, sk))

This was because adjusting alpha manually using the backtracking line search would cause rhok to blow up.  As rhok was a scalar always multiplied by alpha and both were changing, I figured having the code adjust alpha alone would sort it out (which in the end it does).  However I made a lot of adjustments since then and haven't tried adding rhok back in yet to see if the issue was caused by something else.  I'll test that out now